clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,232 @@
+// the tag to the "map" and saving the new "map" to the DB
+func PutTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot modify another user's tags"),
+		}
+	}
+
+	var properties gomatrix.TagProperties
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var tagContent gomatrix.TagContent
+	if len(data) > 0 {
+		if err = json.Unmarshal(data[0].Content, &tagContent); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	} else {
+		tagContent = newTag()
+	}
+	tagContent.Tags[tag] = properties
+	if err = saveTagData(req, localpart, roomID, accountDB, tagContent); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	go func() {
+		if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+			logrus.WithError(err).Error("Failed to send m.tag account data update to syncapi")
+		}
+	}()
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+// Delete functionality works by obtaining the saved tags, removing the intended tag from
+// the "map" and then saving the new "map" in the DB
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot modify another user's tags"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// If there are no tags in the database, exit
+	if len(data) == 0 {
+		// Spec only defines 200 responses for this endpoint so we don't return anything else.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+
+	tagContent := newTag()
+	err = json.Unmarshal(data[0].Content, &tagContent)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// Check whether the tag to be deleted exists
+	if _, ok := tagContent.Tags[tag]; ok {
+		delete(tagContent.Tags, tag)
+	} else {
+		// Spec only defines 200 responses for this endpoint so we don't return anything else.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+	if err = saveTagData(req, localpart, roomID, accountDB, tagContent); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	go func() {
```

Comment:
```suggestion
	// Send data to syncProducer in order to inform clients of changes
	// Run in a goroutine in order to prevent blocking the tag request response
	go func() {
```